File: GILua/src/main/java/org/anime_game_servers/gi_lua/models/PositionImpl.java
package org.anime_game_servers.gi_lua.models;

import lombok.Getter;
import lombok.Setter;

import java.io.Serializable;
import java.util.List;

@Getter
public class PositionImpl implements Serializable, Position {
    @Setter private float x;
    @Setter private float y;
    @Setter private float z;

    public PositionImpl() {}

    public PositionImpl(float x, float y) {
        this.x = x;
        this.y = y;
    }

    public PositionImpl(float x, float y, float z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public PositionImpl(List<Float> xyz) {
        switch (xyz.size()) {
            default:  // Might want to error on excess elements, but maybe we want to extend to 3+3 representation later.
            case 3:
                this.z = xyz.get(2);  // Fall-through
            case 2:
                this.y = xyz.get(1);  // Fall-through
            case 1:
                this.x = xyz.get(0);  // pointless fall-through
            case 0:
                break;
        }
    }
}

File: GILua/src/main/java/org/anime_game_servers/gi_lua/models/PositionImpl.kt
package org.anime_game_servers.gi_lua.models

import java.io.Serializable


data class PositionImpl @JvmOverloads constructor(
    @JvmField var x: Float = 0f,
    @JvmField var y: Float = 0f,
    @JvmField var z:Float = 0f)
    : Serializable, Position {

    override fun getX() = x
    override fun getY() = y
    override fun getZ() = z

    constructor(xyz: MutableList<Float?>) : this(
        xyz.getOrNull(0)?: 0f,
        xyz.getOrNull(1)?: 0f,
        xyz.getOrNull(2)?: 0f
    )
}
